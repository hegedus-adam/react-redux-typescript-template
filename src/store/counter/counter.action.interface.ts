import { actions } from './counter.constant';
import { IState } from './state.interface';

export type AddCounterAction = {
  type: actions.ADD
  payload: {
    value: number
  }
};

export type SubtractCounterAction = {
  type: actions.SUBTRACT
  payload: {
    value: number
  }
};

export type AddCounterActionCreator = (value: number) => AddCounterAction;
export type AddCounterReducer = (state: IState, action: AddCounterAction) => IState;

export type SubtractCounterActionCreator = (value: number) => SubtractCounterAction;
export type SubtractCounterReducer = (state: IState, action: SubtractCounterAction) => IState;

export type CounterActions = AddCounterAction | SubtractCounterAction;
export type CounterActionCreators = AddCounterActionCreator | SubtractCounterActionCreator;
export type CounterReducers = AddCounterReducer | SubtractCounterReducer;
