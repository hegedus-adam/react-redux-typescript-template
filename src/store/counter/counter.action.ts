import { AddCounterActionCreator, SubtractCounterActionCreator } from './counter.action.interface';
import { actions } from './counter.constant';

export const addCounter: AddCounterActionCreator = value => ({
  type: actions.ADD,
  payload: {
    value,
  },
});

export const subtractCounter: SubtractCounterActionCreator = value => ({
  type: actions.SUBTRACT,
  payload: {
    value,
  },
});
