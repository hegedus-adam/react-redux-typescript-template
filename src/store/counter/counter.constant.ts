export enum actions {
  ADD = '@counter/ADD',
  SUBTRACT = '@counter/SUBTRACT',
}
