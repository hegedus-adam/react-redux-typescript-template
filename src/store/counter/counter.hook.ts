import { useSelector } from 'react-redux';
import { IState } from './state.interface';

type CounterSelector = (state: { counter: IState }) => number;
export const selector: CounterSelector = ({ counter: { value } }) => value;

export const useCounter = () => useSelector(selector);
