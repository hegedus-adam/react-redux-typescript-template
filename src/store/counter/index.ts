export { counter } from './counter.reducer';
export { addCounter, subtractCounter } from './counter.action';
export { useCounter } from './counter.hook';
