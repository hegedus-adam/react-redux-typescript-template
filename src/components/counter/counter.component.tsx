import React from 'react';
import { useDispatch } from 'react-redux';
import { addCounter, subtractCounter, useCounter } from '@/store/counter';

export function Counter() {
  const dispatch = useDispatch();
  const counter = useCounter();

  const handleOnAddClick = () => {
    dispatch(addCounter(1));
  };
  const handleOnSubstractClick = () => {
    dispatch(subtractCounter(1));
  };

  return (
    <section>
      <h2>Counter component</h2>
      <p>{counter}</p>
      <button type="button" onClick={handleOnAddClick}>+1</button>
      <button type="button" onClick={handleOnSubstractClick}>-1</button>
    </section>
  );
}
