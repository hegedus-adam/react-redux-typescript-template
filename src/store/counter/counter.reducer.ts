import { reducerFactory } from '@/store/reducer-factory';
import { actions } from './counter.constant';
import { AddCounterReducer, SubtractCounterReducer, CounterReducers } from './counter.action.interface';
import { IState } from './state.interface';

const defaultState: IState = {
  value: 0,
};

const handleAddCounter: AddCounterReducer = (state, { payload: { value } }) => ({
  ...state,
  value: state.value + value,
});

const handleSubtractCounter: SubtractCounterReducer = (state, { payload: { value } }) => ({
  ...state,
  value: state.value - value,
});

const counterHandlers: Record<actions, CounterReducers> = {
  [actions.ADD]: handleAddCounter,
  [actions.SUBTRACT]: handleSubtractCounter,
};

export const counter = reducerFactory(defaultState, counterHandlers);
