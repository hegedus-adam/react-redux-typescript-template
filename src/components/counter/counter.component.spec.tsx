import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import { Counter } from './counter.component';

const mockDispatch = jest.fn();
const mockUseCounter = jest.fn();
const mockUseDispatch = jest.fn(() => mockDispatch);

const mockAddCounter = jest.fn();
const mockSubtractCounter = jest.fn();

jest.mock('react-redux', () => ({
  useDispatch: () => mockUseDispatch(),
}));

jest.mock('@/store/counter', () => ({
  useCounter: () => mockUseCounter(),
  addCounter: (value: number) => mockAddCounter(value),
  subtractCounter: (value: number) => mockSubtractCounter(value),
}));

describe('Counter', () => {
  it('should render component', () => {
    mockUseCounter.mockReturnValueOnce(0);
    const { container } = render(<Counter />);

    expect(container).toMatchSnapshot();
    expect(mockUseCounter.mock.calls.length).toEqual(1);
    expect(mockUseDispatch.mock.calls.length).toEqual(1);
  });

  it('should render component when +1 is clicked', async () => {
    mockUseCounter.mockReturnValueOnce(0);

    render(<Counter />);

    fireEvent.click(screen.getByText('+1'));

    expect(mockUseCounter.mock.calls.length).toEqual(1);
    expect(mockUseDispatch.mock.calls.length).toEqual(1);
    expect(mockDispatch.mock.calls.length).toEqual(1);
    expect(mockAddCounter.mock.calls.length).toEqual(1);
    expect(mockAddCounter.mock.calls[0][0]).toEqual(1);
    expect(mockSubtractCounter.mock.calls.length).toEqual(0);
  });

  it('should render component when -1 is clicked', async () => {
    mockUseCounter.mockReturnValueOnce(0);

    render(<Counter />);

    fireEvent.click(screen.getByText('-1'));

    expect(mockUseCounter.mock.calls.length).toEqual(1);
    expect(mockUseDispatch.mock.calls.length).toEqual(1);
    expect(mockDispatch.mock.calls.length).toEqual(1);
    expect(mockSubtractCounter.mock.calls.length).toEqual(1);
    expect(mockSubtractCounter.mock.calls[0][0]).toEqual(1);
    expect(mockAddCounter.mock.calls.length).toEqual(0);
  });
});
