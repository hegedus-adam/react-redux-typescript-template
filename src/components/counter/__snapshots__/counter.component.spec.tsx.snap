// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should render component 1`] = `
<div>
  <section>
    <h2>
      Counter component
    </h2>
    <p>
      0
    </p>
    <button
      type="button"
    >
      +1
    </button>
    <button
      type="button"
    >
      -1
    </button>
  </section>
</div>
`;
